core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+        self.clf = None
+        self.token_to_id = None
+        self.T = None
+        self.K = None
+        self.occurrence = None
+        self.class_to_answer_group_mapping = None
+        self.top = None
+        self.count_vector = None
+
+    def to_dict(self):
+        """Returns a dict representing this classifier.
+
+        Returns:
+            dict. A dictionary representation of classifier referred as
+            'classifier_data'. This data is used for prediction.
+        """
+        fingerprint_data = {
+            i: {
+                'fingerprint': self.training_data[i]['fingerprint'],
+                'class': self.training_data[i]['class']
+            } for i in self.training_data
+        }
+
+        classifier_data = {
+            'KNN': {
+                'token_to_id': self.token_to_id,
+                'T': self.T,
+                'K': self.K,
+                'top': self.top,
+                'occurrence': self.occurrence,
+                'fingerprint_data': fingerprint_data
+            },
+            'SVM': self.clf.__dict__,
+            'class_to_answer_group_mapping': self.class_to_answer_group_mapping,
+            'cv_vocabulary': self.count_vector.__dict__['vocabulary_']
+        }
+        return classifier_data
+
+    # pylint: disable=too-many-locals
+    def train(self, training_data):
+        """Trains classifier using given training_data.
+
+        Args:
+            training_data: list(dict). The training data that is used for
+                training the classifier. The list contains dicts where each dict
+                represents a single training data group, for example:
+                training_data = [
+                    {
+                        'answer_group_index': 1,
+                        'answers': ['a1', 'a2']
+                    },
+                    {
+                        'answer_group_index': 2,
+                        'answers': ['a2', 'a3']
+                    }
+                ]
+        """
+        # Answer group index to classifier class mapping.
+        answer_group_to_class_mapping = {
+            training_data[i]['answer_group_index']: i
+            for i in range(len(training_data))
+        }
+
+        class_to_answer_group_mapping = dict(zip(
+            answer_group_to_class_mapping.values(),
+            answer_group_to_class_mapping.keys()))
+
+        data = {}
+        count = 0
+        for answer_group in training_data:
+            answer_group_index = answer_group['answer_group_index']
+            for answer in answer_group['answers']:
+                data[count] = {
+                    'source': answer,
+                    'class': answer_group_to_class_mapping[answer_group_index]
+                }
+                count += 1
+
+        data, token_to_id = tokenize_data(data)
+
+        # No. of nearest neighbours to consider for classification using KNN.
+        top = len(answer_group_to_class_mapping) / 2
```

Comment:
Yep that is the K. Actually accuracy decreases 2% when K=2 it works best when K=3. And since we don't have any other dataset to try out this logic, I don't know which will work best.